pkg/services/sqlstore/database_wrapper.go

```diff
@@ -0,0 +1,111 @@
+		Help:      "The total amount of Database queries",
+	}, []string{"status"})
+
+	databaseQueryHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "grafana",
+		Name:      "database_queries_duration_seconds",
+		Help:      "Database query histogram",
+		Buckets:   prometheus.ExponentialBuckets(0.0001, 4, 9),
+	}, []string{"status"})
+
+	prometheus.MustRegister(databaseQueryCounter, databaseQueryHistogram)
+
+	logger = log.New("DBMetrics")
+}
+
+// WrapDatabaseDriverWithHooks creates a fake database driver that
+// executes Pre and post functions which we use to gather metrics about
+// database queries.
+func WrapDatabaseDriverWithHooks(dbType string) string {
+	var d driver.Driver
+
+	drivers := map[string]driver.Driver{
+		migrator.SQLITE:   &sqlite3.SQLiteDriver{},
+		migrator.MYSQL:    &mysql.MySQLDriver{},
+		migrator.POSTGRES: &pq.Driver{},
+	}
+
+	d, exist := drivers[dbType]
+	if !exist {
+		return dbType
+	}
+
+	driverWithHooks := dbType + "WithHooks"
+	sql.Register(driverWithHooks, sqlhooks.Wrap(d, &databaseQueryWrapper{}))
+	core.RegisterDriver(driverWithHooks, &databaseQueryWrapperParser{dbType: dbType})
+	return driverWithHooks
+}
+
+// databaseQueryWrapper satisfies the sqlhook.databaseQueryWrapper interface
+// which allow use to wrapp all SQL queries with a `Before` & `After` hook.
+type databaseQueryWrapper struct{}
+
+// databaseQueryWrapperKey is used as key to save values in `context.Context`
+type databaseQueryWrapperKey struct{}
+
+// Before hook will print the query with it's args and return the context with the timestamp
+func (h *databaseQueryWrapper) Before(ctx context.Context, query string, args ...interface{}) (context.Context, error) {
+	return context.WithValue(ctx, databaseQueryWrapperKey{}, time.Now()), nil
+}
+
+// After hook will get the timestamp registered on the Before hook and print the elapsed time
+func (h *databaseQueryWrapper) After(ctx context.Context, query string, args ...interface{}) (context.Context, error) {
+	begin := ctx.Value(databaseQueryWrapperKey{}).(time.Time)
+	databaseQueryCounter.WithLabelValues("success").Inc()
+	databaseQueryHistogram.WithLabelValues("success").Observe(time.Since(begin).Seconds())
+	logger.Debug("query finished", "status", "success", "elapsed time", time.Since(begin), "sql", query)
+	return ctx, nil
+}
+
+// OnError instances will be called if any error happens
+func (h *databaseQueryWrapper) OnError(ctx context.Context, err error, query string, args ...interface{}) error {
+	status := "error"
+	// https://golang.org/pkg/database/sql/driver/#ErrSkip
+	if err == nil || err == driver.ErrSkip {
+		status = "success"
+	}
+
+	begin := ctx.Value(databaseQueryWrapperKey{}).(time.Time)
+	databaseQueryCounter.WithLabelValues(status).Inc()
+	databaseQueryHistogram.WithLabelValues(status).Observe(time.Since(begin).Seconds())
+	logger.Debug("query finished", "status", status, "elapsed time", time.Since(begin), "sql", query, "error", err)
```

Comment:
I think it's best to call `time.Since(begin)` once and assign to a variable here:
```suggestion
	elapsed := time.Since(begin)
	databaseQueryHistogram.WithLabelValues(status).Observe(elapsed.Seconds())
	logger.Debug("query finished", "status", status, "elapsed time", elapsed, "sql", query, "error", err)
```

Reply:
:+1: